butte-build/src/codegen.rs

```diff
@@ -0,0 +1,559 @@
+                        }
+                    })
+                } else {
+                    None
+                }
+            } else {
+                None
+            }
+        });
+
+        let builder_field_methods = fields.iter().map(|field| {
+            let Field {
+                name: field_name,
+                ty,
+                scalar,
+                ..
+            } = field;
+            let add_method_name = format_ident!("ident_{}", field_name.0);
+            let offset = format_ident!("VT_{}", field_name.0.to_uppercase());
+            let field_offset = quote!(Self::#offset);
+            let body = if ty.is_scalar() {
+                if let Some(default_value) = scalar {
+                    quote!(self.fbb.push_slot(#field_offset, #field_name, #default_value))
+                } else {
+                    quote!(self.fbb.push_slot(#field_offset, #field_name))
+                }
+            } else {
+                quote!(self.fbb.push_slot(#field_offset, #field_name))
+            };
+            quote! {
+                #[inline]
+                fn #add_method_name(&mut self, #field_name: #ty) {
+                    #body;
+                }
+            }
+        });
+        let field_offset_constants = fields.iter().enumerate().map(|(i, Field { name, .. })| {
+            let offset_name = format_ident!("VT_{}", name.0);
+            quote! {
+                pub const #offset_name: flatbuffers::VOffsetT = #i;
+            }
+        });
+        let struct_offset_enum_name = format_ident!("{}Offset", struct_name.0);
+        (quote! {
+            pub enum #struct_offset_enum_name {}
+
+            #[derive(Copy, Clone, Debug, PartialEq, derive_more::From)]
+            #doc
+            pub struct #struct_name {
+                table: flatbuffers::Table<'a>,
+            }
+
+            impl<'a> #struct_name<'a> {
+                pub fn create<'bldr: 'args, 'args: 'mut_bldr, 'mut_bldr>(
+                    fbb: &'mut_bldr mut flatbuffers::FlatBufferBuilder<'bldr>,
+                    args: &'args #args<'args>)
+                    -> flatbuffers::WIPOffset<Self<'bldr>> {
+                    let mut builder = <#builder>::new(fbb);
+                    #(#builder_add_calls);*
+                    builder.finish()
+                }
+
+                // field offset constants
+                #(#field_offset_constants)*
+
+                // nested flatbuffers if applicable
+                #(#field_nested_flatbuffers)*
+            }
+
+            impl<'a> flatbuffers::Follow<'a> for #struct_name<'a> {
+                type Inner = Self;
+
+                #[inline]
+                fn follow(buf: &'a [u8], loc: usize) -> Self::Inner {
+                    let table = flatbuffers::Table { buf, loc };
+                    Self { table }
+                }
+            }
+
+            // Args
+            #[derive(SmartDefault)]
```

Comment:
wtf is this?

Reply:
Potentially a thing I want to not use. I was trying to avoid generating `Default` `impl`s, and found the [`smart-default`](https://github.com/idanarye/rust-smart-default) crate. If there's something better (or you think we should just generate the `Default` `impl`s then I'm all ears).

It's pretty much a band-aid until https://github.com/Centril/rfcs/pull/19 is implemented.

Reply:
Oh, neat! :)